sklearn/ensemble/_weight_boosting.py

```diff
@@ -137,8 +137,13 @@ def fit(self, X, y, sample_weight=None):
         # Initializion of the random number instance that will be used to
         # generate a seed at each iteration
         random_state = check_random_state(self.random_state)
+        epsilon = np.finfo(np.float64).eps
 
         for iboost in range(self.n_estimators):
+            # avoid extremely small sample weight
+            # detail see issue #20320
+            sample_weight = np.clip(sample_weight, a_min=epsilon, a_max=1.0)
```

Comment:
Why clipping above?

```suggestion
            sample_weight = np.clip(sample_weight, a_min=epsilon, a_max=None)
```

also wouldn't it make more sense to set small sample weights to exactly 0 instead (assuming the rest of the code can deal with such exact 0 weights correctly)?

Reply:
1. I set `a_max` to 1.0 because we normalize the sample_weight at each boosting round by `sample_weight /= sample_weight_sum`, so the maximum sample weight would not exceed 1. I agree that setting `a_max` to 1.0 is not needed.

2. Setting small sample weights to 0 hurts model performance which is shown as follows, I guess the reason is those small weights could add up to be impactful at certain node, so they're not ignorable.

AUC under original implementation:
![image](https://user-images.githubusercontent.com/16646940/123725605-ed362080-d8c0-11eb-9911-852a58e7238d.png)

AUC under proposed fix:
![image](https://user-images.githubusercontent.com/16646940/123725698-1a82ce80-d8c1-11eb-99e2-636fb6e3ceae.png)

AUC when setting weight straight to 0:
![image](https://user-images.githubusercontent.com/16646940/123725747-38503380-d8c1-11eb-90be-3668f890eb97.png)